File: LeetCode/0001-two-sum.cpp
struct Number {
    int val;
    int pos;
};

bool CmpNumbersByValue(const Number& a, const Number& b) {
    return a.val <  b.val;
}

const Number& FindNumberByValue(const vector<Number>& sorted_nums, int value) {
    int l = 0, r = sorted_nums.size() - 1;
    while (l < r) {
        int c = (l + r) / 2;
        if (sorted_nums[c].val < value) l = c + 1; else r = c;
    }
    return sorted_nums[l];
}

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();

        // Sort array of nums
        vector<Number> sorted_nums;
        for (int i = 0; i < n; i++) sorted_nums.push_back(Number{nums[i], i});
        sort(sorted_nums.begin(), sorted_nums.end(), &CmpNumbersByValue);

        // Loop through the sorted nums and use bin-search to find a pair
        for (int i = 0; i < n; i++) {
            auto a = sorted_nums[i];
            auto b = FindNumberByValue(sorted_nums, target - a.val);

            // A hack to avoid using the same number twice.
            // We will still find the solution later!
            if (a.pos == b.pos) continue;

            if (a.val + b.val == target) return vector<int>{a.pos, b.pos};
        }

        throw "Solution not exist!";
    }
};

File: LeetCode/0001-two-sum.rs
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut nums: Vec<(usize, i32)> = nums.into_iter().enumerate().collect();
        nums.sort_by(|&(_, a), &(_, b)| a.cmp(&b));
        let mut i = 0usize;
        let mut j = nums.len() - 1;
        loop {
            if i >= j { panic!("Solution must exist!"); }
            let sum = nums[i].1 + nums[j].1;
            if sum == target { break; }
            if sum < target { i += 1; } else { j -= 1; }
        }
        vec![nums[i].0 as i32, nums[j].0 as i32]
    }
}
